src/qibolab/instruments/zhinst.py

```diff
@@ -0,0 +1,1050 @@
+
+        dimensions = []
+        if options.averaging_mode is AveragingMode.SINGLESHOT:
+            dimensions = [options.nshots]
+
+        for sweeper in sweepers:
+            dimensions.append(len(sweeper.values))
+
+        # Re-arranging sweepers based on hardware limitations
+        # FIXME: Punchout and frequency case
+        rearranging_axes = [[], []]
+        if len(sweepers) == 2:
+            if sweepers[1].parameter is Parameter.frequency:
+                if sweepers[0].parameter is Parameter.bias:
+                    rearranging_axes[0] += [sweepers.index(sweepers[1])]
+                    rearranging_axes[1] += [0]
+                    sweeper_changed = sweepers[1]
+                    sweepers.remove(sweeper_changed)
+                    sweepers.insert(0, sweeper_changed)
+                    warnings.warn("Sweepers were reordered")
+                elif (
+                    not sweepers[0].parameter is Parameter.amplitude
+                    and sweepers[0].pulses.type is not PulseType.READOUT
+                ):
+                    rearranging_axes[0] += [sweepers.index(sweepers[1])]
+                    rearranging_axes[1] += [0]
+                    sweeper_changed = sweepers[1]
+                    sweepers.remove(sweeper_changed)
+                    sweepers.insert(0, sweeper_changed)
+                    warnings.warn("Sweepers were reordered")
+
+        # TODO: Read frequency for pulses instead of qubit patch
+        for qubit in qubits.values():
+            for pulse in sequence:
+                if pulse.qubit == qubit.name:
+                    if pulse.type is PulseType.READOUT:
+                        qubit.readout_frequency = pulse.frequency
+                    if pulse.type is PulseType.DRIVE:
+                        qubit.drive_frequency = pulse.frequency
+        """
+        Play pulse sequence steps, one on each method:
+        Translation, Calibration, Experiment Definition and Execution.
+        """
+        self.sweepers = sweepers
+        self.sequence_zh(sequence, qubits, sweepers)
+        self.calibration_step(qubits)
+        self.create_exp(qubits, options)
+        self.run_exp()
+
+        # TODO: General, several readouts and qubits
+        # TODO: Implement the new results!
+        "Get the results back"
+        results = {}
+        for qubit in qubits.values():
+            if not qubit.flux_coupler:
+                if self.sequence[f"readout{qubit.name}"]:
+                    exp_res = self.results.get_data(f"sequence{qubit.name}")
+                    # Reorder dimensions
+                    exp_res = np.moveaxis(exp_res, rearranging_axes[0], rearranging_axes[1])
+                    if options.acquisition_type is AcquisitionType.DISCRIMINATION:
+                        if options.averaging_mode is AveragingMode.CYCLIC:
+                            data = np.array([exp_res])
+                        else:
+                            data = np.array(exp_res)
```

Comment:
Something extremely similar to this was also above, maybe you could reuse the same function.